modules/base/tool_test.go

```diff
@@ -0,0 +1,160 @@
+package base
+
+import (
+	"testing"
+
+	"github.com/go-gitea/gitea/modules/setting"
+	"github.com/stretchr/testify/assert"
+	"strk.kbt.io/projects/go/libravatar"
+)
+
+func TestEncodeMD5(t *testing.T) {
+	assert.Equal(t,
+		"3858f62230ac3c915f300c664312c63f",
+		EncodeMD5("foobar"),
+	)
+}
+
+func TestEncodeSha1(t *testing.T) {
+	assert.Equal(t,
+		"8843d7f92416211de9ebb963ff4ce28125932878",
+		EncodeSha1("foobar"),
+	)
+}
+
+func TestShortSha(t *testing.T) {
+	assert.Equal(t, "veryverylo", ShortSha("veryverylong"))
+}
+
+// TODO: Test DetectEncoding()
+
+func TestBasicAuthDecode(t *testing.T) {
+	_, _, err := BasicAuthDecode("?")
+	assert.Equal(t, "illegal base64 data at input byte 0", err.Error())
+
+	user, pass, err := BasicAuthDecode("Zm9vOmJhcg==")
+	assert.NoError(t, err)
+	assert.Equal(t, "foo", user)
+	assert.Equal(t, "bar", pass)
+}
+
+func TestBasicAuthEncode(t *testing.T) {
+	assert.Equal(t, "Zm9vOmJhcg==", BasicAuthEncode("foo", "bar"))
+}
+
+func TestGetRandomString(t *testing.T) {
+	assert.Len(t, GetRandomString(4), 4)
+}
+
+// TODO: Test PBKDF2()
+// TODO: Test VerifyTimeLimitCode()
+// TODO: Test CreateTimeLimitCode()
+
+func TestHashEmail(t *testing.T) {
+	assert.Equal(t,
+		"d41d8cd98f00b204e9800998ecf8427e",
+		HashEmail(""),
+	)
+	assert.Equal(t,
+		"353cbad9b58e69c96154ad99f92bedc7",
+		HashEmail("gitea@example.com"),
+	)
+}
+
+func TestAvatarLink(t *testing.T) {
+	setting.EnableFederatedAvatar = false
+	setting.LibravatarService = nil
+	setting.DisableGravatar = true
+
+	assert.Equal(t, "/img/avatar_default.png", AvatarLink(""))
+
+	setting.DisableGravatar = false
+	assert.Equal(t,
+		"353cbad9b58e69c96154ad99f92bedc7",
+		AvatarLink("gitea@example.com"),
+	)
+
+	setting.EnableFederatedAvatar = true
+	assert.Equal(t,
+		"353cbad9b58e69c96154ad99f92bedc7",
+		AvatarLink("gitea@example.com"),
+	)
+	setting.LibravatarService = libravatar.New()
+	assert.Equal(t,
+		"http://cdn.libravatar.org/avatar/353cbad9b58e69c96154ad99f92bedc7",
+		AvatarLink("gitea@example.com"),
```

Comment:
well, this one isn't really testing the *federated* part (I've seen packages just prefixing the hash with "cdn.libravatar.org" and be happy with it, it's not our case, we're actually doing the DNS lookups)